// tml/Gruntfile.js
module.exports = function(grunt) {
    var matchdep = require('matchdep'); // dependencies from package

    var ptcSourceFiles = [
        'src/vuforia-angular.js',
        'src/loader/**/*.js',
        'test/cases/**/*.js',
        'test/helpers/**/*.js'
    ];
    var cfg = {
        downloadDeps:  [
            {
                src: 'https://cdnjs.cloudflare.com/ajax/libs/ionic/1.2.4/js/ionic.bundle.min.js',
                dest: 'test/libs/ionic.bundle.min.js',
                hash: '6b5bd38deea6292c29eb1669a8f8bdf3d6ffc1ea87760ffa6ea62e11f99179a2'
            },
            {
                src: 'https://ajax.googleapis.com/ajax/libs/angularjs/1.4.3/angular-mocks.js',
                dest: 'test/libs/angular-mocks.js',
                hash: 'f9f1cbc6b899dbd102660fef7aa47eb6c63607f4bc78d1fed2e259d35398108a'
            }
        ],

        karma: {
            unit: {
                configFile: 'karma.conf.js',
                singleRun: true
            }
        },

        lintspaces: {
            sources: ['src/**/*.js', 'test/cases/**/*.js', 'test/helpers/**/*.js'],
            options: {
                endOfLine: 'lf'
            }
        },
        jscs: {
            src: ptcSourceFiles,
            options: {
                config: ".jscsrc"
            }
        },
        jshint: {
            src: ptcSourceFiles
        }
    };

    matchdep.filterDev(['grunt-*', '!grunt-cli']).forEach(grunt.loadNpmTasks);
    grunt.loadTasks('grunt-tasks/');
    grunt.initConfig(cfg);

    grunt.registerTask('default', [
        'all'
    ]);

    grunt.registerTask('test', [
        'downloadDeps',
        'karma'
    ]);

    grunt.registerTask('lint', [
        'lintspaces',
        'jscs',
        'jshint'
    ]);

    grunt.registerTask('all', [
        'lint',
        'test'
    ]);
};
